File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        int[] visited = new int[numCourses];

        for (int[] req : prerequisites) {
            int course = req[0];
            int prereq = req[1];
            map.putIfAbsent(course, new ArrayList<>());
            map.get(course).add(prereq);
        }

        for (int i = 0; i < numCourses; i++) {
            if (visited[i] == 0 && !dfs(i, map, visited)) {
                return false;
            }
        }

        return true;
    }

    private boolean dfs(int course, Map<Integer, List<Integer>> map, int[] visited) {
        if (visited[course] == 1) {
            return false;
        }

        if (visited[course] == 2) {
            return true;
        }

        visited[course] = 1;

        if (map.containsKey(course)) {
            for (int prereq : map.get(course)) {
                if (!dfs(prereq, map, visited)) {
                    return false;
                }
            }
        }

        visited[course] = 2;

        return true;
    }
}

File: 0207-course-schedule/0207-course-schedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = {}
        visited = [0] * numCourses

        for course, prereq in prerequisites:
            if course not in graph:
                graph[course] = []
            graph[course].append(prereq)

        def dfs(course):
            if visited[course] == 1:
                return False
            if visited[course] == 2:
                return True

            visited[course] = 1

            if course in graph:
                for prereq in graph[course]:
                    if not dfs(prereq):
                        return False

            visited[course] = 2
            return True

        for i in range(numCourses):
            if visited[i] == 0 and not dfs(i):
                return False

        return True
